### ISSUE TITLE
PYQT5+YOLO

### ISSUE THREAD

**Issue by yqsl123** [ultralytics/ultralytics#18120]
### Search before asking

- [X] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I want to use PYQT to design a GUI to detect images, videos, and cameras. And the ability to change the confidence and IoU. The category and number of detected targets are then displayed.How can I improve this code to implement the above features?
'''
from PyQt5.QtWidgets import QApplication, QMainWindow, QFileDialog
import sys

from PyQt5 import QtCore, QtGui, QtWidgets
import torch
import torchvision
import cv2
from ultralytics import YOLO

class Ui_MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setupUi(self)
        self.retranslateUi(self)

    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(1131, 932)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.pushButton = QtWidgets.QPushButton(self.centralwidget)
        self.pushButton.setGeometry(QtCore.QRect(40, 20, 98, 26))
        self.pushButton.setObjectName("pushButton")
        self.pushButton_2 = QtWidgets.QPushButton(self.centralwidget)
        self.pushButton_2.setGeometry(QtCore.QRect(220, 20, 98, 26))
        self.pushButton_2.setObjectName("pushButton_2")
        self.pushButton_3 = QtWidgets.QPushButton(self.centralwidget)
        self.pushButton_3.setGeometry(QtCore.QRect(390, 20, 98, 26))
        self.pushButton_3.setObjectName("pushButton_3")
        self.textBrowser = QtWidgets.QTextBrowser(self.centralwidget)
        self.textBrowser.setGeometry(QtCore.QRect(40, 70, 1031, 61))
        self.textBrowser.setObjectName("textBrowser")
        self.label = QtWidgets.QLabel(self.centralwidget)
        self.label.setGeometry(QtCore.QRect(40, 180, 511, 451))
        self.label.setObjectName("label")
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        self.label_2.setGeometry(QtCore.QRect(590, 180, 491, 461))
        self.label_2.setObjectName("label_2")
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 1131, 23))
        self.menubar.setObjectName("menubar")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

        self.pushButton.clicked.connect(self.uploadImage)
        self.pushButton_2.clicked.connect(self.showEnvironment)
        self.pushButton_3.clicked.connect(self.startProgram)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.pushButton.setText(_translate("MainWindow", "Upload an image"))

        self.pushButton_2.setText(_translate("MainWindow", "Display the environment"))
        self.pushButton_3.setText(_translate("MainWindow", "Start the program"))
        self.label.setText(_translate("MainWindow", "             Original image"))
        self.label_2.setText(_translate("MainWindow", "             Test results"))

    def uploadImage(self):
        file_dialog = QFileDialog()
        image_path, _ = file_dialog.getOpenFileName(self, 'Select an image', '', 'Images (*.png *.xpm *.jpg *.bmp)')
        self.image_path = image_path
        if image_path:
        
            pixmap = QtGui.QPixmap(image_path)
            self.label.setPixmap(pixmap)
            self.label.setScaledContents(True)

    def showEnvironment(self):
        pytorch_version = torch.__version__
        torchvision_version = torchvision.__version__
        self.textBrowser.setText(f"PyTorch Version: {pytorch_version}\n"
                                 f"Torchvision Version: {torchvision_version}")

    def startProgram(self):
        self.textBrowser.setText(self.image_path)
        model = YOLO(r'yolov8n.pt')
        results = model(self.image_path)
        annotated_frame = results[0].plot()

        save_image_name = self.image_path.split("/")[-1].split(".")[0] + "_pred"
        cv2.imwrite(f"C:\work\PycharmProjects\BscanObjDetection\data\predict\{save_image_name}.jpg", annotated_frame)

    
        height, width, channel = annotated_frame.shape
        bytes_per_line = 3 * width
        qimage = QtGui.QImage(annotated_frame.data, width, height, bytes_per_line, QtGui.QImage.Format_RGB888)
 
        pixmap = QtGui.QPixmap.fromImage(qimage)

        self.label_2.setPixmap(pixmap)
        self.label_2.setScaledContents(True)

if __name__ == '__main__':
    app = QApplication(sys.argv)
    MainWindow1 = QMainWindow()  
    ui = Ui_MainWindow()  
    ui.setupUi(MainWindow1)
    MainWindow1.show()
    sys.exit(app.exec_())
'''

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yqsl123, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users, where you can find various [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many of the most common questions may already be addressed there.

If this is a ❓ Question or request for guidance (e.g., improving your PYQT5+YOLO integration), please provide as much information about your use case as possible, including any specific challenges you're running into or areas where you're looking for improvement. The code and context you've supplied are helpful, thank you for including them! 😊

For issues involving 🐛 bugs, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us accurately debug any problems.

We also recommend the following steps to troubleshoot your implementation:
- Confirm you're using the latest version of the `ultralytics` package. You can upgrade it with:
```bash
pip install -U ultralytics
```
- Read through [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and verify your setup aligns with our best practices.
- Test your setup in one of our verified environments listed in the [Environments guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/).

If you'd like to connect with the community for further discussions or collaboration, consider joining our [Discord](https://discord.com/invite/ultralytics) 🎧, [Discourse](https://community.ultralytics.com), or [Subreddit](https://reddit.com/r/Ultralytics).

## Resources for Real-Time Testing

To quickly test and interact with YOLO models in a GUI-based environment, consider using one of our preconfigured environments:
- Launch a notebook with free GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

## Status and Updates
You can check if all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing by looking at this badge:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
These tests ensure all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) are functioning on macOS, Windows, and Ubuntu.

This is an automated response; an Ultralytics engineer will review your inquiry and provide further assistance soon. Thank you for your patience! 😊

**Comment by glenn-jocher** [comment]
@yqsl123 your current implementation is a good start! To incorporate confidence and IoU threshold adjustments as well as display the detected categories and counts, you can extend the `startProgram` method. Add sliders in your GUI for these thresholds, and pass their values to the `model.predict` method. Use the `results.boxes` attribute to extract class names and counts. Refer to the [Gradio + YOLO11 integration example](https://docs.ultralytics.com/integrations/gradio/) for ideas on handling these parameters dynamically. Let us know if you need further guidance!

**Comment by glenn-jocher** [comment]
@yqsl123 to enhance your existing code, you can add sliders for confidence and IoU thresholds and display detected category names, confidence levels, and coordinates in the `startProgram` method. Use `QtWidgets.QSlider` for dynamic threshold adjustment and extract detection data using `results[0].boxes`. Here's an example of accessing and displaying detection information:

```python
results = model(self.image_path, conf=confidence_threshold, iou=iou_threshold)
for box in results[0].boxes:
    category = results[0].names[int(box.cls)]
    confidence = box.conf
    coordinates = box.xyxy
    self.textBrowser.append(f"Category: {category}, Confidence: {confidence:.2f}, Coordinates: {coordinates.tolist()}")
```

Integrate this logic into your GUI for interactivity. For further guidance, refer to the [Ultralytics documentation](https://docs.ultralytics.com/) or reach out with specific questions!
